Allow setting on hardware backed signals

DominicOram opened the pull request:
Fixes #1077

### Instructions to reviewer on how to test:
1. Confirm new tests pass

### Checks for reviewer
- [ ] Would the PR title make sense to a scientist on a set of release notes
- [ ] If a new device has been added does it follow the [standards](https://diamondlightsource.github.io/dodal/main/reference/device-standards.html)
- [ ] If changing the API for a pre-existing device, ensure that any beamlines using this device have updated their Bluesky plans accordingly
- [ ] Have the connection tests for the relevant beamline(s) been run via `dodal connect ${BEAMLINE}`

DominicOram commented on src/dodal/common/signal_utils.py:
```diff
@@ -2,22 +2,24 @@
     def __init__(
         self,
         get_from_hardware_func: Callable[[], Coroutine[Any, Any, SignalDatatypeT]],
+        set_to_hardware_func: Callable[[], Coroutine[SignalDatatypeT, Any, Any]] = None,
```
Thanks! I missed it 'cos the linter is still broken due to https://github.com/DiamondLightSource/dodal/pull/1072

Relm-Arrowny commented on src/dodal/common/signal_utils.py:
```diff
@@ -27,6 +31,40 @@ async def get_value(self) -> SignalDatatypeT:
+            return await super().put(value, wait)
+
+
+def create_rw_hardware_backed_soft_signal(
+    datatype: type[SignalDatatypeT],
+    get_from_hardware_func: Callable[[], Coroutine[Any, Any, SignalDatatypeT]],
+    set_to_hardware_func: SetHardwareType,
+    units: str | None = None,
+    precision: int | None = None,
+):
```
Should: I think we should pass the initial_value as we are using it in put. same for the create_hardware_backed_soft_signal
```suggestion
def create_rw_hardware_backed_soft_signal(
    datatype: type[SignalDatatypeT],
    get_from_hardware_func: Callable[[], Coroutine[Any, Any, SignalDatatypeT]],
    set_to_hardware_func: Callable[[SignalDatatypeT], Coroutine[Any, Any, None]],
    initial_value: SignalDatatypeT | None = None,
    units: str | None = None,
    precision: int | None = None,
) -> SignalRW:
```

DominicOram replied:
Yh, I think it's an edge case anyway and `initial_value` is probably never going to be `None`

Relm-Arrowny commented on src/dodal/common/signal_utils.py:
```diff
@@ -27,6 +31,40 @@ async def get_value(self) -> SignalDatatypeT:
+    )
+
 
 def create_hardware_backed_soft_signal(
```
Could: Make it match soft_r /soft_rw ?
```suggestion
def create_r_hardware_backed_soft_signal(
```

Relm-Arrowny reviewed (commented):
As I try to use it, I spotted a couple of things.

olliesilvester commented on src/dodal/common/signal_utils.py:
```diff
@@ -27,8 +31,42 @@ async def get_value(self) -> SignalDatatypeT:
+        if self.set_to_hardware_func:
+            write_value = self.initial_value if value is None else value
+            await self.set_to_hardware_func(write_value)
+        else:
+            return await super().put(value, wait)
```
Do we need this else statement? If `set_to_hardware_func` is set to None, then this should be the backend of a `SignalR` and so `set_value` can't ever be called

olliesilvester reviewed (changes requested):
Looks good, just a few minor comments and questions

The device standards page in the docs refer to `create_hardware_backed_soft_signal`, so we should change that too

DominicOram commented:
I've addressed some comments but given https://github.com/bluesky/ophyd-async/pull/661 is close to complete and will replace this I don't want to spend too much more time on it.
